nexus/db-queries/src/db/datastore/ip_pool.rs

```diff
@@ -324,6 +334,80 @@ impl DataStore {
             })
     }
 
+    pub async fn ip_pool_allocated_count(
+        &self,
+        opctx: &OpContext,
+        authz_pool: &authz::IpPool,
+    ) -> Result<IpsAllocated, Error> {
+        opctx.authorize(authz::Action::Read, authz_pool).await?;
+
+        use db::schema::external_ip;
+        use diesel::dsl::sql;
+        use diesel::sql_types::BigInt;
+
+        let (ipv4, ipv6) = external_ip::table
+            .filter(external_ip::ip_pool_id.eq(authz_pool.id()))
+            .filter(external_ip::time_deleted.is_null())
+            .select((
+                sql::<BigInt>("count(*) FILTER (WHERE family(ip) = 4)"),
+                sql::<BigInt>("count(*) FILTER (WHERE family(ip) = 6)"),
+            ))
+            .first_async::<(i64, i64)>(
+                &*self.pool_connection_authorized(opctx).await?,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+
+        Ok(IpsAllocated { ipv4, ipv6 })
+    }
+
+    pub async fn ip_pool_total_capacity(
+        &self,
+        opctx: &OpContext,
+        authz_pool: &authz::IpPool,
+    ) -> Result<IpsCapacity, Error> {
+        opctx.authorize(authz::Action::Read, authz_pool).await?;
+        opctx.authorize(authz::Action::ListChildren, authz_pool).await?;
+
+        use db::schema::ip_pool_range;
+
+        let ranges = ip_pool_range::table
+            .filter(ip_pool_range::ip_pool_id.eq(authz_pool.id()))
+            .filter(ip_pool_range::time_deleted.is_null())
+            .select(IpPoolRange::as_select())
+            // This is a rare unpaginated DB query, which means we are
+            // vulnerable to a resource exhaustion attack in which someone
+            // creates a very large number of ranges in order to make this
+            // query slow. In order to mitigate that, we limit the query to the
+            // (current) max allowed page size, effectively making this query
+            // exactly as vulnerable as if it were paginated. If there are more
+            // than 10,000 ranges in a pool, we will undercount, but I have a
+            // hard time seeing that as a practical problem.
+            .limit(10000)
```

Comment:
In [private chat](https://matrix.to/#/!YNYPOVxjAUeXksTcRj:oxide.computer/$NV_J25yvwwrIByPZw0QFaS7-Hl2AA3TfmSfkKwNNnfc?via=oxide.computer&via=unix.house&via=matrix.org) we discussed the alternative of doing this calculation in SQL so we don't have to limit it here. What we found is that neither CRDB nor Postgres support subtraction of IPv6 addresses when the result is greater than a 64 bit integer, and we couldn't figure out how to convert the addresses to 128 bit integers first and then subtract. So it seems we are forced to do this in Rust.

```
> select 'ffff::'::inet - '::1';
ERROR: result out of range
```

https://github.com/cockroachdb/cockroach/blob/91c1d5cbb5a56cd9c082d78814cd03c5471dddfa/pkg/util/ipaddr/ipaddr_test.go#L862-L881

https://github.com/cockroachdb/cockroach/blob/4deb9e3b77f38396a58bb8186fe51bb01cf9f007/pkg/util/ipaddr/ipaddr.go#L607-L632